server/src/main/java/org/elasticsearch/cluster/routing/IndexRoutingTable.java

```diff
@@ -328,7 +326,7 @@ public static Builder builder(Index index) {
     public static class Builder {
 
         private final Index index;
-        private final Map<Integer, IndexShardRoutingTable> shards = new HashMap<>();
+        private Object[] shards;
```

Comment:
Could we make this a `IndexShardRoutingTable.Builder[]` if we made deserialize not use the builder? The other usage I found was `RoutingTable.updateNumberOfReplicas`, but that seems to create builders anyway.

That would avoid the instanceof and casting introduced here.

Reply:
Good point and this turned out to be really trivial to do actually (apart from some diff noise in tests) because the deserialization was creating builders anyway.
=> done :)